// coreui/lib/components/Form/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Trigger = exports.Summary = exports.Message = exports.Field = exports.Context = exports.Button = exports.addInputTypes = exports.default = undefined;

var _Form = require('./Form');

Object.defineProperty(exports, 'addInputTypes', {
  enumerable: true,
  get: function get() {
    return _Form.addInputTypes;
  }
});
Object.defineProperty(exports, 'Button', {
  enumerable: true,
  get: function get() {
    return _Form.Button;
  }
});
Object.defineProperty(exports, 'Context', {
  enumerable: true,
  get: function get() {
    return _Form.Context;
  }
});
Object.defineProperty(exports, 'Field', {
  enumerable: true,
  get: function get() {
    return _Form.Field;
  }
});
Object.defineProperty(exports, 'Message', {
  enumerable: true,
  get: function get() {
    return _Form.Message;
  }
});
Object.defineProperty(exports, 'Summary', {
  enumerable: true,
  get: function get() {
    return _Form.Summary;
  }
});
Object.defineProperty(exports, 'Trigger', {
  enumerable: true,
  get: function get() {
    return _Form.Trigger;
  }
});

var _Form2 = _interopRequireDefault(_Form);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = _Form2.default;
