// react-semantic-form/es6/Field.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

import React from 'react';
import Fieldset from './Fieldset';
import * as inputs from './inputs';
import { findDOMNode } from 'react-dom';

class Field extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      error: false
    };
  }

  componentDidMount() {
    const name = this.props.name;
    this.context.fieldset.registerInput(name, this);
  }

  componentWillUnmount() {
    const name = this.props.name;
    this.context.fieldset.unregisterInput(name, this);
  }

  getValue() {
    return this.refs.inp.getValue();
  }

  validate(value) {
    return this.refs.inp.validate(value).then(value => {
      this.setState({
        error: false
      });
      return value;
    }).catch(err => {
      console.log("Error ", err);
      this.setState({
        error: err.toString()
      });
    });
  }

  setFocus() {
    const inp = this.refs.inp;
    if (inp.setFocus) {
      inp.setFocus();
    } else {
      const node = findDOMNode(inp);
      if (node && node.focus) {
        node.focus();
      }
    }
  }

  render() {
    const { label, className, type, name, ...other } = this.props;
    const error = this.state.error;
    const Input = typeof type === "string" ? inputs.hasOwnProperty(type) ? inputs[type] : inputs.Text : type;
    const defaultValue = this.context.fieldset.getDefaultValue(name);
    return React.createElement(
      'div',
      { className: "inline fields ui grid" },
      React.createElement(
        'div',
        { className: "four wide column field" + (error ? " error" : "") },
        label && React.createElement(
          'label',
          null,
          label
        )
      ),
      React.createElement(
        'div',
        { className: "twelve wide column field" + (error ? " error" : "") },
        React.createElement(Input, _extends({ ref: 'inp', name: name }, other, { value: defaultValue })),
        typeof error === "string" && React.createElement(
          'p',
          null,
          error
        )
      )
    );
  }
}

Field.contextTypes = {
  fieldset: React.PropTypes.object
};

//
// Field.connect = function(component) {
//   return React.createClass({
//     contextTypes: {
//       fieldset: React.PropTypes.instanceOf(Fieldset)
//     },
//
//     componentDidMount() {
//       const name = this.props.name;
//       this.context.fieldset.registerInput(name, this.refs.inp);
//     },
//
//     componentWillUnmount() {
//       const name = this.props.name;
//       this.context.fieldset.unregisterInput(name, this.refs.inp);
//     },
//
//     render() {
//       return <component ref="inp" {...this.props} />
//     }
//
//   });
// }

export default Field;
