// coreui/lib/components/Table/TableHelpers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Table = require('../../theme/components/Table');

var _Table2 = _interopRequireDefault(_Table);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _ramda = require('ramda');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var currentPage = function currentPage(pagination, pageIndex, pageSize, data) {
  return !pagination ? data : (0, _ramda.splitEvery)(pageSize, data)[pageIndex] || [];
};

var isColumnSearchable = function isColumnSearchable(columnId, searchable) {
  return (0, _ramda.is)(Array, searchable) ? new Set(searchable).has(columnId) : !!searchable;
};

var isColumnSortable = function isColumnSortable(columnId, sortable) {
  return (0, _ramda.is)(Array, sortable) ? new Set(sortable).has(columnId) : !!sortable;
};

var isKeyColumnValid = function isKeyColumnValid(data, valueField) {
  return !!(valueField && new Set(data.filter(function (r) {
    return !(0, _ramda.isNil)(r[valueField]);
  })).size === data.length);
};

var isSearchMatchingRow = function isSearchMatchingRow(columns, searchValue, row) {
  return (0, _ramda.any)(function (c) {
    return c.isSearchable && (row[c.id] || '').toString().toLowerCase().includes(searchValue.toLowerCase());
  }, columns);
};

var maxPageIndex = function maxPageIndex(data, pageSize) {
  return (0, _ramda.dec)((0, _ramda.splitEvery)(pageSize, data).length);
};

var normalizedColumns = function normalizedColumns(baseTableProps, columns) {
  var data = baseTableProps.data;
  var helpers = baseTableProps.helpers;
  var searchable = baseTableProps.searchable;
  var sortable = baseTableProps.sortable;


  return (columns || (0, _ramda.uniq)((0, _ramda.chain)(_ramda.keys, data))).map(function (c) {
    var component = c.component;
    var displayName = (0, _ramda.isNil)(c.displayName) ? c : c.displayName;
    var id = c.id || c;
    var isSearchable = helpers.isColumnSearchable(id, searchable);
    var isSortable = helpers.isColumnSortable(id, sortable);

    return { component: component, displayName: displayName, id: id, isSearchable: isSearchable, isSortable: isSortable };
  });
};

var normalizedProps = function normalizedProps(helpers, props) {
  var pagination = props.pagination;
  var pageIndex = props.pageIndex;
  var pageSize = props.pageSize;
  var prevPageIndex = props.prevPageIndex;
  var searchable = props.searchable;
  var selection = props.selection;
  var searchValue = props.searchValue;
  var sortable = props.sortable;
  var sortAscending = props.sortAscending;
  var sortField = props.sortField;
  var valueField = props.valueField;


  var columns = helpers.normalizedColumns({ data: props.data, helpers: helpers, searchable: searchable, sortable: sortable }, props.columns);

  var sortedData = helpers.sortedData({ columns: columns, searchable: searchable, searchValue: searchValue, sortAscending: sortAscending, sortField: sortField }, helpers, props.data);

  var data = helpers.currentPage(pagination, (0, _ramda.isNil)(pageIndex) ? prevPageIndex : pageIndex, pageSize, sortedData);

  return (0, _ramda.merge)(props, {
    columns: columns,
    data: data,
    maxPageIndex: helpers.maxPageIndex(sortedData, pageSize),
    searchable: !!searchable,
    selection: !!(selection && helpers.isKeyColumnValid(data, valueField)),
    sortable: !!sortable
  });
};

var sortedData = function sortedData(baseTableProps, helpers, data) {
  var columns = baseTableProps.columns;
  var searchable = baseTableProps.searchable;
  var searchValue = baseTableProps.searchValue;
  var sortAscending = baseTableProps.sortAscending;
  var sortField = baseTableProps.sortField;

  var filteredData = searchable ? data.filter((0, _ramda.partial)(helpers.isSearchMatchingRow, [columns, searchValue])) : data;
  var xs = !sortField ? filteredData : (0, _ramda.sortBy)((0, _ramda.prop)(sortField), filteredData);

  return sortField && sortAscending === false ? (0, _ramda.reverse)(xs) : xs;
};

var classes = _Table2.default.classes;
var options = _Table2.default.options;
var styles = _Table2.default.styles;

var tableDefaultProps = function tableDefaultProps() {
  return {
    columns: null,
    data: [],
    pageSize: 10,
    pagination: false,
    searchable: false,
    searchPlaceholder: 'Search items...',
    selectMultiple: true,
    selection: true,
    sortable: true,
    sortAscending: true,
    theme: { classes: classes, options: options, styles: styles }
  };
};

var toggleRow = function toggleRow(selectedRows, rowId) {
  if (selectedRows.has(rowId)) {
    selectedRows.delete(rowId);
  } else {
    selectedRows.add(rowId);
  }

  return selectedRows;
};

exports.default = {
  currentPage: currentPage, isColumnSearchable: isColumnSearchable, isColumnSortable: isColumnSortable, isKeyColumnValid: isKeyColumnValid, isSearchMatchingRow: isSearchMatchingRow,
  maxPageIndex: maxPageIndex, normalizedColumns: normalizedColumns, normalizedProps: normalizedProps, sortedData: sortedData, tableDefaultProps: tableDefaultProps, toggleRow: toggleRow
};
