// coreui/lib/Shared.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _jss = require('jss');

var _jss2 = _interopRequireDefault(_jss);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _jssExtend = require('jss-extend');

var _jssExtend2 = _interopRequireDefault(_jssExtend);

var _jssNested = require('jss-nested');

var _jssNested2 = _interopRequireDefault(_jssNested);

var _jssCamelCase = require('jss-camel-case');

var _jssCamelCase2 = _interopRequireDefault(_jssCamelCase);

var _jssDefaultUnit = require('jss-default-unit');

var _jssDefaultUnit2 = _interopRequireDefault(_jssDefaultUnit);

var _jssVendorPrefixer = require('jss-vendor-prefixer');

var _jssVendorPrefixer2 = _interopRequireDefault(_jssVendorPrefixer);

var _reactWidgets = require('./jss/react-widgets');

var _reactWidgets2 = _interopRequireDefault(_reactWidgets);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Shared = function Shared() {
  var _this = this;

  _classCallCheck(this, Shared);

  this.attachReactWidgetsSheet = function () {
    _this.jss.createStyleSheet(_reactWidgets2.default, { named: false }).attach();
  };

  this.getRegisteredComponents = function () {
    return _this.importedComponents;
  };

  this.registerComponent = function (name, component) {
    _this.importedComponents[name] = component;
  };

  this.importedComponents = {};
  this.jss = _jss2.default.create();
  this.useSheet = (0, _reactJss2.default)(this.jss);
  this.jss.use((0, _jssExtend2.default)());
  this.jss.use((0, _jssNested2.default)());
  this.jss.use((0, _jssCamelCase2.default)());
  this.jss.use((0, _jssDefaultUnit2.default)());
  this.jss.use((0, _jssVendorPrefixer2.default)());
  this.attachReactWidgetsSheet();
};

var SharedInst = new Shared();

exports.default = SharedInst;
